rust/protocol/src/curve.rs

```diff
@@ -1,55 +1,80 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
-mod curve25519;
+//! Identity creation primitives.
 
+#![deny(missing_docs)]
+
+pub mod curve25519;
+
+use curve25519::{PRIVATE_KEY_LENGTH, PUBLIC_KEY_LENGTH};
+
+use crate::utils::constant_time_cmp;
 use crate::{Result, SignalProtocolError};
 
 use std::cmp::Ordering;
 use std::convert::TryFrom;
 use std::fmt;
 
 use arrayref::array_ref;
+use num_enum::{IntoPrimitive, TryFromPrimitive, TryFromPrimitiveError};
 use rand::{CryptoRng, Rng};
 use subtle::ConstantTimeEq;
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Map the variant of key being used to and from a [u8].
+///
+/// Implements [SignalProtocolError::from] on top of [num_enum::TryFromPrimitive] to convert into
+/// raising a [SignalProtocolError] if the encoded key variant is invalid.
+///
+///```
+/// # fn main() -> libsignal_protocol::error::Result<()> {
+/// use libsignal_protocol::KeyType;
+/// use std::convert::{TryFrom, TryInto};
+///
+/// let encoded_key_type: u8 = KeyType::Curve25519.into();
+/// assert!(encoded_key_type == 0x05);
+///
+/// let original_key_type: KeyType = encoded_key_type.try_into()?;
+/// assert!(original_key_type == KeyType::Curve25519);
+///
+/// let bad_encoded_key: u8 = 0x27;
+/// assert!(KeyType::try_from(bad_encoded_key).is_err());
+///
+/// # Ok(())
+/// # }
+///```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
 pub enum KeyType {
-    Djb,
+    /// See [Curve25519].
+    ///
+    /// [Curve25519]: https://en.wikipedia.org/wiki/Curve25519.
+    Curve25519 = 0x05,
 }
 
-impl fmt::Display for KeyType {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        fmt::Debug::fmt(self, f)
+impl From<TryFromPrimitiveError<KeyType>> for SignalProtocolError {
+    fn from(err: TryFromPrimitiveError<KeyType>) -> Self {
+        SignalProtocolError::BadKeyType(err.number)
     }
 }
 
-impl KeyType {
-    fn value(&self) -> u8 {
-        match &self {
-            KeyType::Djb => 0x05u8,
-        }
-    }
-}
-
-impl TryFrom<u8> for KeyType {
-    type Error = SignalProtocolError;
-
-    fn try_from(x: u8) -> Result<Self> {
-        match x {
-            0x05u8 => Ok(KeyType::Djb),
-            t => Err(SignalProtocolError::BadKeyType(t)),
-        }
-    }
+/// Interface for structs that perform operations parameterized by values of [KeyType].
+pub trait Keyed {
+    /// Return the variant of key this object employs.
+    fn key_type(&self) -> KeyType;
 }
 
 #[derive(Debug, Clone, Copy, Eq, PartialEq)]
 enum PublicKeyData {
-    DjbPublicKey([u8; 32]),
+    Curve25519PublicKey([u8; PUBLIC_KEY_LENGTH]),
```

Comment:
This `PUBLIC_KEY_LENGTH` would probably look better *with* qualification, so it's obvious where we're getting it from. (If we ever had two types it would certainly be necessary.)

Reply:
Also while you're here we probably don't need to repeat "PublicKey" in the case name.

Reply:
Done and done!